lib/Secrets/secrets_c.h

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2017 Jolla Ltd.
+ * Contact: Chris Adams <chris.adams@jollamobile.com>
+ * All rights reserved.
+ * BSD 3-Clause License, see LICENSE.
+ */
+
+#ifndef LIBSAILFISHSECRETS_SECRETS_C_H
+#define LIBSAILFISHSECRETS_SECRETS_C_H
+
+/* This file provides a C-compatible wrapper for Secrets */
+
+#include <stddef.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+/****************************** result ******************************/
+
+enum Sailfish_Secrets_Result_Code {
+    Sailfish_Secrets_Result_Succeeded = 0,
+    Sailfish_Secrets_Result_Pending = 1,
+    Sailfish_Secrets_Result_Failed = 2
+};
+
+struct Sailfish_Secrets_Result {
+    enum Sailfish_Secrets_Result_Code code;
+    int errorCode;
+    char *errorMessage;
+};
+
+struct Sailfish_Secrets_Result*
+Sailfish_Secrets_Result_new(
+        enum Sailfish_Secrets_Result_Code code,
+        int errorCode,
+        const char *errorMessage);
+
+void Sailfish_Secrets_Result_delete(
+        struct Sailfish_Secrets_Result *result);
+
+/****************************** secret ******************************/
+
+struct Sailfish_Secrets_Secret_Identifier {
+    char *name;
+    char *collectionName;
+};
+
+struct Sailfish_Secrets_Secret_FilterDatum {
+    char *field;
+    char *value;
+    struct Sailfish_Secrets_Secret_FilterDatum *next;
+};
+
+struct Sailfish_Secrets_Secret {
+    struct Sailfish_Secrets_Secret_Identifier *identifier;
+    struct Sailfish_Secrets_Secret_FilterDatum *filterData;
+    unsigned char *data;
+    size_t dataSize;
+};
+
+struct Sailfish_Secrets_Secret_Identifier*
+Sailfish_Secrets_Secret_Identifier_new(
+        const char *name,
+        const char *collectionName);
+
+void Sailfish_Secrets_Secret_Identifier_delete(
+        struct Sailfish_Secrets_Secret_Identifier *ident);
```

Comment:
I would strongly suggest to make objects and any potentially shareable data structures ref-countable (i.e. use new/ref/unref style API). That makes life so much easier and costs so little on the implementation side that I can't think of any reason not to do it.

Reply:
Ok, I'll add the reference count and ref/unref methods, and remove the delete methods

Reply:
Should I just add a (publicly visible) int refcount; member to each struct?  One alternative would be to add a void* member (which I allocate manually, but that guarantees another heap allocation).  A final alternative would be to only provide access/setter functions, and just expose the struct as an opaque handle, but that's suboptimal for a few reasons IMO...

Reply:
You missed another alternative! [Here](https://git.merproject.org/mer-core/libdbusaccess/blob/master/src/dbusaccess_peer.c#L51) is an example. Basically, the real (internal) structure is bigger than the public part, the internal part containing refcount or whatever else you don't want to expose. There are examples of similar technique in glib itself (e.g. GArray). That's what I like the most (except when I need a GObject, then the approach is slightly different).